src/ostree/ot-builtin-rev-parse.c

```diff
@@ -25,47 +25,79 @@
 #include "ot-builtins.h"
 #include "ostree.h"
 #include "otutil.h"
+#include <stdbool.h>
 
 /* ATTENTION:
  * Please remember to update the bash-completion script (bash/ostree) and
  * man page (man/ostree-rev-parse.xml) when changing the option list.
  */
 
+static gboolean opt_single;
+
 static GOptionEntry options[] = {
+  { "single", 'S', 0, G_OPTION_ARG_NONE, &opt_single, "If the repository has exactly one commit, then print it; any other case will result in an error", NULL },
   { NULL }
 };
 
 gboolean
 ostree_builtin_rev_parse (int argc, char **argv, OstreeCommandInvocation *invocation, GCancellable *cancellable, GError **error)
 {
-  g_autoptr(GOptionContext) context = NULL;
+  g_autoptr(GOptionContext) context = g_option_context_new ("REV");
   g_autoptr(OstreeRepo) repo = NULL;
-  gboolean ret = FALSE;
-  const char *rev = "master";
-  int i;
-  g_autofree char *resolved_rev = NULL;
+  if (!ostree_option_context_parse (context, options, &argc, &argv, invocation, &repo, cancellable, error))
+    return FALSE;
 
-  context = g_option_context_new ("REV");
+  if (opt_single)
+    {
+      if (argc >= 2)
+        {
+          ot_util_usage_error (context, "Cannot specify arguments with --single", error);
+          return FALSE;
+        }
 
-  if (!ostree_option_context_parse (context, options, &argc, &argv, invocation, &repo, cancellable, error))
-    goto out;
+      g_autoptr(GHashTable) objects = NULL;
+      if (!ostree_repo_list_commit_objects_starting_with (repo, "", &objects, cancellable, error))
+        return FALSE;
+
+      GHashTableIter iter;
+      void* key = NULL;
+      void *value = NULL;
+      bool found = false;
+      g_hash_table_iter_init (&iter, objects);
+      while (g_hash_table_iter_next (&iter, &key, &value))
+        {
+          if (found)
+            return glnx_throw (error, "Multiple commit objects found");
+          found = true;
+        }
+      if (!found)
+        return glnx_throw (error, "No commit objects found");
+      g_assert (key != NULL);
+      GVariant *serialized_key = key;
+      const char *checksum;
+      OstreeObjectType objtype;
+      ostree_object_name_deserialize (serialized_key, &checksum, &objtype);
```

Comment:
Hmm, why not use `GLNX_HASH_TABLE_FOREACH` here? Ahh because you want the key to live longer.

OK how about:

```suggestion
      GVariant *found = NULL;
      GLNX_HASH_TABLE_FOREACH (objects, GVariant*, key)
        {
          if (found)
            return glnx_throw (error, "Multiple commit objects found");
          found = key;
        }
      if (!found)
        return glnx_throw (error, "No commit objects found");
      const char *checksum;
      OstreeObjectType objtype;
      ostree_object_name_deserialize (found, &checksum, &objtype);
```